File: Motocitizen/src/motocitizen/MyApp.java
package motocitizen;

import android.content.Context;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.support.multidex.MultiDex;
import android.support.multidex.MultiDexApplication;
import android.widget.Toast;

import com.google.firebase.messaging.FirebaseMessaging;
import com.vk.sdk.VKAccessToken;
import com.vk.sdk.VKAccessTokenTracker;
import com.vk.sdk.VKSdk;

import motocitizen.datasources.database.Database;
import motocitizen.datasources.preferences.Preferences;
import motocitizen.geo.MyGoogleApiClient;
import motocitizen.geo.geocoder.MyGeoCoder;
import motocitizen.ui.activity.AuthActivity;
import motocitizen.user.Auth;
import motocitizen.utils.GraphUtils;

public class MyApp extends MultiDexApplication {

    @Override
    protected void attachBaseContext(Context base) {
        super.attachBaseContext(base);
        MultiDex.install(this);
    }

    public static void logoff() {
        Auth.INSTANCE.logoff();
    }

    public static boolean isOnline(Context context) {
        ConnectivityManager cm      = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo         netInfo = cm.getActiveNetworkInfo();
        return netInfo != null && netInfo.isConnectedOrConnecting();
    }

    @Override
    public void onCreate() {
        super.onCreate();
        Preferences.INSTANCE.initialize(this);
        Database.INSTANCE.initialize(this);
        GraphUtils.INSTANCE.initialize(this);
        MyGeoCoder.INSTANCE.initialize(this);
        MyGoogleApiClient.INSTANCE.initialize(this);
        FirebaseMessaging.getInstance().subscribeToTopic("accidents");
        vkAccessTokenTracker.startTracking();
        VKSdk.initialize(this);
    }

    /**
     * AccessToken invalidated. Слушатель токена
     */
    VKAccessTokenTracker vkAccessTokenTracker = new VKAccessTokenTracker() {
        @Override
        public void onVKAccessTokenChanged(VKAccessToken oldToken, VKAccessToken newToken) {
            if (newToken == null) {
                Toast.makeText(getApplicationContext(), "Авторизация слетела, авторизируйтесь снова", Toast.LENGTH_LONG).show();
                Intent intent = new Intent(getApplicationContext(), AuthActivity.class);
                intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
                startActivity(intent);
            }
        }
    };
}

File: Motocitizen/src/motocitizen/MyApp.kt
package motocitizen

import android.annotation.SuppressLint
import android.app.Application
import android.content.Context
import android.content.Intent
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.widget.Toast
import androidx.preference.PreferenceManager
import com.vk.api.sdk.VK
import com.vk.api.sdk.VKTokenExpiredHandler
import motocitizen.datasources.preferences.Preferences
import motocitizen.migration.Migration
import motocitizen.notifications.Messaging
import motocitizen.ui.activity.AuthActivity

class MyApp : Application() {


    /**
     * AccessToken invalidated. Слушатель токена
     */
    private var vkAccessTokenHandler = object : VKTokenExpiredHandler {
        override fun onTokenExpired() {
            Toast.makeText(applicationContext, "Авторизация слетела, авторизируйтесь снова", Toast.LENGTH_LONG).show()
            val intent = Intent(applicationContext, AuthActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
            startActivity(intent)
        }
    }

    override fun onCreate() {
        super.onCreate()
        context = applicationContext
        oldVersion = PreferenceManager.getDefaultSharedPreferences(this).getInt("mc.app.version", 0)
        val currentVersion = packageManager.getPackageInfo(packageName, 0).longVersionCode
        if (oldVersion < currentVersion) Migration.makeMigration(this)

        Messaging.subscribe()
        if (Preferences.isTester) Messaging.subscribeToTest()

        VK.initialize(this)
        VK.addTokenExpiredHandler(vkAccessTokenHandler)
    }

    companion object {
        @SuppressLint("StaticFieldLeak")
        lateinit var context: Context
        var firstStart = false
        var oldVersion = 0

        fun isOnline(context: Context): Boolean {
            val connectivityManager = context.getSystemService(CONNECTIVITY_SERVICE) as ConnectivityManager
            val network = connectivityManager.activeNetwork ?: return false
            val capabilities = connectivityManager.getNetworkCapabilities(network) ?: return false
            return capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) &&
                    capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)
        }
    }
}
